File: 2020/day4.cpp
#include <iostream>
#include <string>
#include <regex>
#include <unordered_set>
#include <memory>

#include "adventofcode.h"

class Passport
{
public:
	int byr, iyr, eyr;
	std::string hgt, hcl, ecl, pid, cid;

	virtual bool validate()
	{
		// ignore empty cid
		if (byr == 0 || iyr == 0 || eyr == 0 || hgt.empty() || hcl.empty() || ecl.empty() || pid.empty())
		{
			return false;
		}
		return true;
	}
};

class Passport2 : public Passport
{
public:
	std::unordered_set<std::string> valid_ecl = {"amb", "blu", "brn", "gry", "grn", "hzl", "oth"};

	virtual bool validate()
	{
		std::size_t pos;
		if (!Passport::validate())
		{
			std::cerr << "Missing fields" << std::endl;
			return false;
		}
		if (byr < 1920 || byr > 2002)
		{
			std::cerr << "Invalid byr: " << byr << std::endl;
			return false;
		}
		if (iyr < 2010 || iyr > 2020)
		{
			std::cerr << "Invalid iyr: " << iyr << std::endl;
			return false;
		}
		if (eyr < 2020 || eyr > 2030)
		{
			std::cerr << "Invalid eyr: " << eyr << std::endl;
			return false;
		}
		if ((pos = hgt.find("cm")) != std::string::npos)
		{
			int h = std::stoi(hgt.substr(0, pos));
			if (h < 150 || h > 193)
			{
				std::cerr << "Invalid hgt (cm): " << h << std::endl;
				return false;
			}
		}
		else if ((pos = hgt.find("in")) != std::string::npos)
		{
			int h = std::stoi(hgt.substr(0, pos));
			if (h < 59 || h > 76)
			{
				std::cerr << "Invalid hgt (in): " << h << std::endl;
				return false;
			}
		}
		else
		{
			return false;
		}
		if (!std::regex_match(hcl, std::regex(R"(^#[\da-f]{6}$)")))
		{
			std::cerr << "Invalid hcl: " << hcl << std::endl;
			return false;
		}
		if (valid_ecl.find(ecl) == valid_ecl.end())
		{
			std::cerr << "Invalid ecl: " << hcl << std::endl;
			return false;
		}
		if (!std::regex_match(pid, std::regex(R"(^\d{9}$)")))
		{
			std::cerr << "Invalid pid: " << pid << std::endl;
			return false;
		}

		return true;
	}
};

template <class T>
void day4(std::istream &file)
{
	int num_valid = 0;
	std::string line;
	auto current = std::unique_ptr<Passport>(new T());
	std::regex reg(R"(([a-z]+):([a-z\d#]+))");
	while (std::getline(file, line))
	{
		// when we get to an empty line, we validate and go to the next passport
		if (line == "")
		{
			if (current->validate())
				num_valid++;
			current.reset(new T());
			continue;
		}

		std::smatch match;
		std::string::const_iterator it(line.begin());
		while (std::regex_search(it, line.cend(), match, reg))
		{
			it = match.suffix().first;

			auto field = match[1].str();
			auto value = match[2].str();
			if (field == "byr")
				current->byr = std::stoi(value);
			else if (field == "iyr")
				current->iyr = std::stoi(value);
			else if (field == "eyr")
				current->eyr = std::stoi(value);
			else if (field == "hgt")
				current->hgt = value;
			else if (field == "hcl")
				current->hcl = value;
			else if (field == "ecl")
				current->ecl = value;
			else if (field == "pid")
				current->pid = value;
			else if (field == "cid")
				current->cid = value;
		}
	}

	if (current->validate())
	{
		num_valid++;
	}

	std::cout << num_valid << std::endl;
}

void day4_1(std::istream &file)
{
	day4<Passport>(file);
}

void day4_2(std::istream &file)
{
	day4<Passport2>(file);
}

File: 2020/rust/src/days/day4.rs
use crate::solver::Solver;
use regex::Regex;

#[derive(Default)]
struct Passport {
    byr: i16,
    iyr: i16,
    eyr: i16,
    hgt: String,
    hcl: String,
    ecl: String,
    pid: String,
    // cid is never used
}

const VALID_ECL: &[&str] = &["amb", "blu", "brn", "gry", "grn", "hzl", "oth"];

impl Passport {
    fn parse(fields: impl Iterator<Item = String>) -> Self {
        let mut passport = Passport::default();
        let field_reg = Regex::new(r"([a-z]+):([a-z\d#]+)").unwrap();
        for ref field in fields {
            if let Some(matches) = field_reg.captures(field) {
                let field = matches.get(1).unwrap().as_str();
                let value = matches.get(2).unwrap().as_str();
                match field {
                    "byr" => passport.byr = value.parse().unwrap_or_default(),
                    "iyr" => passport.iyr = value.parse().unwrap_or_default(),
                    "eyr" => passport.eyr = value.parse().unwrap_or_default(),
                    "hgt" => passport.hgt = value.to_owned(),
                    "hcl" => passport.hcl = value.to_owned(),
                    "ecl" => passport.ecl = value.to_owned(),
                    "pid" => passport.pid = value.to_owned(),
                    _ => (),
                }
            }
        }
        passport
    }

    fn validate_part1(&self) -> bool {
        !(self.byr == 0
            || self.iyr == 0
            || self.eyr == 0
            || self.hgt.is_empty()
            || self.hcl.is_empty()
            || self.ecl.is_empty()
            || self.pid.is_empty())
    }

    fn validate_part2(&self) -> bool {
        if !self.validate_part1() {
            return false;
        }
        if self.byr < 1920 || self.byr > 2002 {
            return false;
        }
        if self.iyr < 2010 || self.iyr > 2020 {
            return false;
        }
        if self.eyr < 2020 || self.eyr > 2030 {
            return false;
        }
        if self.hgt.contains("cm") {
            if let Ok(h) = self.hgt[..self.hgt.len() - 2].parse::<i16>() {
                if !(150..=193).contains(&h) {
                    return false;
                }
            }
        } else if self.hgt.contains("in") {
            if let Ok(h) = self.hgt[..self.hgt.len() - 2].parse::<i16>() {
                if !(59..=76).contains(&h) {
                    return false;
                }
            }
        } else {
            return false;
        }
        if !Regex::new(r"^#[\da-f]{6}$")
            .unwrap()
            .is_match(self.hcl.as_str())
        {
            return false;
        }
        if !VALID_ECL.contains(&self.ecl.as_str()) {
            return false;
        }
        if !Regex::new(r"(^\d{9}$)")
            .unwrap()
            .is_match(self.pid.as_str())
        {
            return false;
        }

        true
    }
}

pub struct Day4 {
    passports: Vec<Passport>,
}

impl Day4 {
    fn solve_part1(&self) -> i64 {
        let mut num_valid = 0;
        for passport in &self.passports {
            if passport.validate_part1() {
                num_valid += 1;
            }
        }
        num_valid
    }

    fn solve_part2(&self) -> i64 {
        let mut num_valid = 0;
        for passport in &self.passports {
            if passport.validate_part2() {
                num_valid += 1;
            }
        }
        num_valid
    }
}

impl Solver for Day4 {
    fn new(lines: impl Iterator<Item = String>) -> Self {
        let mut passports = Vec::new();
        let mut fields = Vec::new();
        for line in lines {
            if line.is_empty() {
                passports.push(Passport::parse(fields.into_iter()));
                fields = Vec::new();
            }
            fields.append(
                line.split(' ')
                    .map(|l| l.to_owned())
                    .collect::<Vec<_>>()
                    .as_mut(),
            );
        }
        passports.push(Passport::parse(fields.into_iter()));
        Day4 { passports }
    }

    fn solve(&mut self) -> (i64, i64) {
        (self.solve_part1(), self.solve_part2())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;

    const INPUT_PART1: &str = indoc! {"
        ecl:gry pid:860033327 eyr:2020 hcl:#fffffd
        byr:1937 iyr:2017 cid:147 hgt:183cm

        iyr:2013 ecl:amb cid:350 eyr:2023 pid:028048884
        hcl:#cfa07d byr:1929

        hcl:#ae17e1 iyr:2013
        eyr:2024
        ecl:brn pid:760753108 byr:1931
        hgt:179cm

        hcl:#cfa07d eyr:2025 pid:166559648
        iyr:2011 ecl:brn hgt:59in
    "};

    #[test]
    fn part1() {
        let day4 = Day4::new(INPUT_PART1.lines().map(|l| l.to_owned()));
        assert_eq!(day4.solve_part1(), 2);
    }

    const INPUT_PART2: &str = indoc! {"
        eyr:1972 cid:100
        hcl:#18171d ecl:amb hgt:170 pid:186cm iyr:2018 byr:1926

        iyr:2019
        hcl:#602927 eyr:1967 hgt:170cm
        ecl:grn pid:012533040 byr:1946

        hcl:dab227 iyr:2012
        ecl:brn hgt:182cm pid:021572410 eyr:2020 byr:1992 cid:277

        hgt:59cm ecl:zzz
        eyr:2038 hcl:74454a iyr:2023
        pid:3556412378 byr:2007

        pid:087499704 hgt:74in ecl:grn iyr:2012 eyr:2030 byr:1980
        hcl:#623a2f

        eyr:2029 ecl:blu cid:129 byr:1989
        iyr:2014 pid:896056539 hcl:#a97842 hgt:165cm

        hcl:#888785
        hgt:164cm byr:2001 iyr:2015 cid:88
        pid:545766238 ecl:hzl
        eyr:2022

        iyr:2010 hgt:158cm hcl:#b6652a ecl:blu byr:1944 eyr:2021 pid:093154719
    "};

    #[test]
    fn part2() {
        let day4 = Day4::new(INPUT_PART2.lines().map(|l| l.to_owned()));
        assert_eq!(day4.solve_part2(), 4);
    }
}
